README.md
pybacklog
=================

Backlog API v2 Client Library for Python

[![Build Status](https://travis-ci.org/netmarkjp/pybacklog.svg?branch=master)](https://travis-ci.org/netmarkjp/pybacklog)

# Requirements

- Python 2.7 or Python 3.5+
- requests 2.x

# Install

```
pip install pybacklog
```

# Usage

```python
from pybacklog import BacklogClient

client = BacklogClient("your_space_name", "your_api_key")

# space
space = client.do("GET", "space")  # GET /api/v2/space
print(space.get(u"spaceKey"))

# project
projects = client.projects()

# activity
activities = client.project_activities("YOUR_PROJECT", {"activityTypeId[]": [1, 2]})

# list issue
project_id = client.get_project_id("YOUR_PROJECT")
issues = client.issues({"projectId[]":[project_id], "sort": "dueDate"})

# specified issue
issue = client.issue("YOUR_PROJECT-999")

# create issue
project_id = client.get_project_id(project_key)
issue_type_id = client.project_issue_types(project_key)[0][u"id"]
priority_id = client.priorities()[0][u"id"]

client.create_issue(project_id,
                    u"some summary",
                    issue_type_id,
                    priority_id,
                    {"description": u"a is b and c or d."})

# add comment
client.add_issue_comment("YOUR_PROJECT-999", u"or ... else e.")

# top 10 star collector
star_collectors = [(client.user_stars_count(u[u"id"], {"since": "2017-06-01", "until": "2017-06-30"})[u"count"], u[u"name"]) for u in client.users()]
star_collectors.sort()
star_collectors.reverse()

for i, (c, u) in enumerate(star_collectors[:10]):
    print(i+1, c, u)
```

supported operations are `pydoc pybacklog.BacklogClient`

extra parameters are here
=> [Backlog API Overview \| Backlog Developer API \| Nulab](https://developer.nulab-inc.com/docs/backlog/)

# Unsupported operations ?

Use `do` or let's write code and Pull Request.

```python
from pybacklog import BacklogClient

client = BacklogClient("your_space_name", "your_api_key")
space = client.do("GET", "space")  # GET /api/v2/space
projects = client.do("GET", "projects",
          query_params={"archived": false}
          )  # GET /api/v2/projects?archived=false
activities = client.do("GET", "projects/{project_id_or_key}/activities",
          url_params={"project_id_or_key": "myproj"},
          query_params={"activityTypeId[]": [1, 2]}
          )  # GET /api/v2/projects/myproj/activities?activityTypeIds%5B%5D=1&activityTypeIds%5B%5D=2
```

see also [Backlog API Overview \| Backlog Developer API \| Nulab](https://developer.nulab-inc.com/docs/backlog/)

# Development

```
pip install -r requirements.txt
pip install -r requirements_dev.txt

PYTHONPATH=. python -m unittest tests
```

# License

Copyright 2017 Toshiaki Baba

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.


CHANGELOG.md
### 0.1.8 (Tue Sep 8 15:11:32 2020 +0900)

- Merge pull request #13 from netmarkjp/tidying-requrements (Tue Aug 27 10:37:17 2019 +0900) [78099cb](https://github.com/netmarkjp/pybacklog/commit/78099cbd70c4997829f714dff82285eaa0fd8169)
- Merge pull request #1 from netmarkjp/master (Mon Jul 20 23:26:19 2020 +0900) [44c35ee](https://github.com/netmarkjp/pybacklog/commit/44c35eee5a59b3bfe178661e43c39e45c9518c2c)
- Merge pull request #14 from tys-hiroshi/master (Tue Sep 8 15:11:32 2020 +0900) [85e8dc7](https://github.com/netmarkjp/pybacklog/commit/85e8dc7d2d405e248160a1e5da732a1171dda213)

### 0.1.7 (Tue Dec 11 11:32:58 2018 +0900)

- Merge pull request #12 from tys-hiroshi/master (Tue Dec 11 11:32:58 2018 +0900) [3c87f3c](https://github.com/netmarkjp/pybacklog/commit/3c87f3cd801ce764459bc2ad1588512ddfffb573)

### 0.1.6 (Mon Nov 12 11:10:57 2018 +0900)

- Merge pull request #10 from tys-hiroshi/master (Mon Nov 12 11:10:57 2018 +0900) [72b659c](https://github.com/netmarkjp/pybacklog/commit/72b659c80aec543f819ee64e9df73946d586d733)

### 0.1.5 (Thu Nov 8 16:03:31 2018 +0900)

- Merge pull request #7 from hirosassa/milestone (Tue Jan 23 18:39:29 2018 +0900) [441d51c](https://github.com/netmarkjp/pybacklog/commit/441d51c234a2201f584701d6fbfe1fb9b9c0fcfd)
- Merge pull request #9 from netmarkjp/auto-detect-domain-com-jp (Thu Nov 8 16:03:31 2018 +0900) [5c1bbab](https://github.com/netmarkjp/pybacklog/commit/5c1bbabe4db44ed348cfbb2b31bfedf9047626fa)

### 0.1.3 (Mon Jan 15 22:38:42 2018 +0900)

- Merge branch 'pr-5' (Mon Dec 25 15:01:18 2017 +0900) [52215fe](https://github.com/netmarkjp/pybacklog/commit/52215fec6e5a3ab2c9ae8fc47931dc3423083d73)
- Merge pull request #4 from hirosassa/add-wiki-methods (Tue Dec 26 09:30:48 2017 +0900) [f6a0632](https://github.com/netmarkjp/pybacklog/commit/f6a0632ecc91e9275db2c12ba7d93a0a97329d66)
- Merge pull request #6 from hirosassa/project_users (Mon Jan 15 22:36:12 2018 +0900) [d9062e4](https://github.com/netmarkjp/pybacklog/commit/d9062e4772d93b07aafe45e7847ccda6cf7316f1)

### 0.1.1 (Thu Dec 7 18:22:26 2017 +0900)

- Merge pull request #3 from yukikaoru/fix-requires-on-setup (Thu Dec 7 18:20:18 2017 +0900) [a759c0c](https://github.com/netmarkjp/pybacklog/commit/a759c0c80d82559a0f4cee4619c40cf6f19ea43f)

### 0.1.0 (Fri Jun 23 12:20:46 2017 +0900)

- Merge pull request #1 from netmarkjp/fix-invalid-character (Wed Jun 21 19:06:17 2017 +0900) [b74db4f](https://github.com/netmarkjp/pybacklog/commit/b74db4f2bf03c18fb45e0a1957eb62cfd49d85a0)
- Merge pull request #2 from netmarkjp/add-utility-methods (Fri Jun 23 12:20:46 2017 +0900) [eaa09ec](https://github.com/netmarkjp/pybacklog/commit/eaa09eccf770a13a17c76616b5f3e61710996c37)


### 0.0.1 (Wed Jan 4 01:51:27 2017 +0900)

- add about install (Wed Jan 4 01:51:27 2017 +0900) [2db65d4](https://github.com/netmarkjp/pybacklog/commit/2db65d4d22aeac9c029351e64b3a91dba04ef889)


tests/__init__.py
# -*- coding: utf-8 -*-

from pybacklog import BacklogClient
import unittest


class TestBacklogClient(unittest.TestCase):

    def test_init(self):
        try:
            client = BacklogClient("my_space_name", "my_api_key")
            self.fail()
        except Exception as _ex:
            self.assertEqual(str(_ex), "retrive space information failed. maybe space not found in .com nor .jp")

    def test_remove_mb4(self):
        testing = (
            (
                {"equal1": u"あいう", "equal2": u"123１２３"},
                {"equal1": u"あいう", "equal2": u"123１２３"}
            ),
            (
                {"replaced1": u"あい💔", "replaced2": u"123♥２３"},
                {"replaced1": u"あい\uFFFD", "replaced2": u"123♥２３"}
            ),
        )
        for t in testing:
            self.assertEqual(BacklogClient.remove_mb4(t[0]), t[1])


examples/project_activities.py
# -*- coding: utf-8 -*-

from pybacklog import BacklogClient
import os

_space = os.getenv("BACKLOG_SPACE")
_api_key = os.getenv("BACKLOG_API_KEY")
_project = os.getenv("BACKLOG_PROJECT")


client = BacklogClient(_space, _api_key)
activities = client.do("GET", "projects/{project_id_or_key}/activities",
                       url_params={"project_id_or_key":    _project},
                       query_params={"activityTypeId[]": [
                           1, 2, 3, 14], "count": 100}
                       )

urls = []
items = []
for activity in activities:
    url = client.activity_to_issue_url(activity)
    if url in urls:
        continue
    urls.append(url)

    item = (activity.get(u"created"), url,
            activity.get(u"content").get(u"summary"))
    items.append(item)

for item in items:
    print(u"{date}\t{url}\t{summary}".format(date=item[0], url=item[1], summary=item[2]))


examples/project_activities2.py
# -*- coding: utf-8 -*-

from pybacklog import BacklogClient
import os

_space = os.getenv("BACKLOG_SPACE")
_api_key = os.getenv("BACKLOG_API_KEY")
_project = os.getenv("BACKLOG_PROJECT")


client = BacklogClient(_space, _api_key)
activities = client.project_activities(
    _project,
    {"activityTypeId[]": [1, 2, 3, 14], "count": 100})

urls = []
items = []
for activity in activities:
    url = client.activity_to_issue_url(activity)
    if url in urls:
        continue
    if "None" in url:
        print(activity)
        continue
    urls.append(url)

    item = (activity.get(u"created"), url,
            activity.get(u"content").get(u"summary"))
    items.append(item)

for item in items:
    print(u"{date}\t{url}\t{summary}".format(
        date=item[0], url=item[1], summary=item[2]))


pybacklog/__init__.py
# -*- coding: utf-8 -*-

import requests
import re


class BacklogClient(object):

    def __init__(self, space_name, api_key):
        self.space_name = space_name
        self.api_key = api_key

        ## auto detetcion of space location
        self.endpoint = BacklogClient._detect_endpoint(space_name, api_key)


    @staticmethod
    def _detect_endpoint(space_name, api_key):
        # at first try .com (new default)
        _endpoint = "https://%s.backlog.com/api/v2/{path}" % space_name
        resp = requests.get(_endpoint.format(path="space"), params={"apiKey": api_key})
        if resp.status_code == 401:
            # space found but got 401 (Authentication failure)
            raise Exception(resp, resp.text)
        try:
            space_key = resp.json().get("spaceKey")
            if space_key == space_name:
                return _endpoint
        except Exception:
            # space not found
            pass

        # if space not found in .com, try .jp
        _endpoint = "https://%s.backlog.jp/api/v2/{path}" % space_name
        resp = requests.get(_endpoint.format(path="space"), params={"apiKey": api_key})
        if resp.status_code == 401:
            # space found but got 401 (Authentication failure)
            raise Exception(resp, resp.text)
        try:
            space_key = resp.json().get("spaceKey")
            if space_key == space_name:
                return _endpoint
        except Exception:
            # space not found
            pass

        raise Exception("retrive space information failed. maybe space not found in .com nor .jp")


    def do(self, method, url, url_params={}, query_params={}, request_params={}):
        """
        - Method: method
        - URL: url.format(**url_params)
        - Parameter: query_params & apiKey=api_key
        - Request Body(data): request_params
        """
        _url = url.format(**url_params).lstrip("/")
        _endpoint = self.endpoint.format(path=_url)
        _headers = {"Content-Type": "application/x-www-form-urlencoded"}

        request_params = BacklogClient.remove_mb4(request_params)

        resp = None

        method = method.lower().strip()
        query_params.update({"apiKey": self.api_key})
        if method == "get":
            resp = requests.get(_endpoint, params=query_params)
        elif method == "patch":
            resp = requests.patch(
                _endpoint, params=query_params, data=request_params, headers=_headers)
        elif method == "post":
            resp = requests.post(
                _endpoint, params=query_params, data=request_params, headers=_headers)
        elif method == "delete":
            resp = requests.delete(
                _endpoint, params=query_params, data=request_params, headers=_headers)
        else:
            raise Exception("Unsupported Method")

        if resp.status_code >= 400:
            raise Exception(resp, resp.text)

        if resp.status_code == 204:
            # 204 NO_CONTENT is blank response
            # used in star
            return None

        return resp.json()

    def activity_to_issue_url(self, activity):
        url = "https://{space}.backlog.jp/view/{project_key}-{content_id}".format(
            space=self.space_name,
            project_key=activity.get(u"project").get(u"projectKey"),
            content_id=activity.get(u"content").get(u"key_id"),
        )
        return url

    @staticmethod
    def remove_mb4(request_params):
        # remove 4 byte characters
        pattern = re.compile(u"[^\u0000-\uD7FF\uE000-\uFFFF]", re.UNICODE)
        for key in request_params.keys():
            try:
                if isinstance(request_params[key], unicode):
                    request_params[key] = pattern.sub(
                        u"\uFFFD", request_params[key])
            except NameError:
                # maybe python3
                request_params[key] = pattern.sub(
                    u"\uFFFD", request_params[key])
        return request_params

    # -------------------------------
    # major operations (PR welcome)
    # -------------------------------
    #
    # - required values => args
    # - optional values => extra_query_params, extra_request_params
    #     - url_params may always required

    def space(self):
        """
        client = BacklogClient("your_space_name", "your_api_key")
        client.space()
        """
        return self.do("GET", "space")

    def projects(self, extra_query_params={}):
        """
        client = BacklogClient("your_space_name", "your_api_key")
        client.projects()
        client.projects({"archived": "false",})
        """
        return self.do("GET", "projects", query_params=extra_query_params)

    def project_activities(self, project_id_or_key, extra_query_params={}):
        """
        client = BacklogClient("your_space_name", "your_api_key")
        client.project_activities("YOUR_PROJECT")
        client.project_activities("YOUR_PROJECT", {"activityTypeId[]": [1, 2],})
        """
        return self.do("get", "projects/{project_id_or_key}/activities",
                       url_params={"project_id_or_key": project_id_or_key},
                       query_params=extra_query_params,
                       )

    def project_users(self, project_id_or_key):
        """
        client = BacklogClient("your_space_name", "your_api_key")
        client.project_users("YOUR_PROJECT")
        """
        return self.do("GET", "projects/{project_id_or_key}/users",
                       url_params={"project_id_or_key": project_id_or_key},
                       )

    def versions(self, project_id_or_key):
        """
        client = BacklogClient("your_space_name", "your_api_key")
        client.versions(3)
        """
        return self.do("GET", "projects/{project_id_or_key}/versions",
                       url_params={"project_id_or_key": project_id_or_key},
                       )

    def create_version(self, project_id_or_key, version_name, extra_request_params={}):
        """
        client = BacklogClient("your_space_name", "your_api_key")

        client.create_version("YOUR_PROJECT",
                              "VERSION_NAME",
                              {"description": "version description"})
        """
        request_params = extra_request_params
        request_params["name"] = version_name
        return self.do("POST", "projects/{project_id_or_key}/versions",
                       url_params={"project_id_or_key": project_id_or_key},
                       request_params=request_params,
                       )

    def update_version(self, project_id_or_key, version_id, version_name, extra_request_params={}):        
        """
        client = BacklogClient("your_space_name", "your_api_key")
        client.update_version("YOUR_PROJECT",
                              3,
                              "VERSION_NAME"
                              {"description": "updated description",
                               "archived": "true"})
        """
        
        request_params = extra_request_params
        request_params["name"] = version_name
        return self.do("PATCH", "projects/{project_id_or_key}/versions/{version_id}",
                       url_params={"project_id_or_key": project_id_or_key,
                                   "version_id": version_id},
                       request_params=request_params,
                       )

    def delete_version(self, project_id_or_key, version_id):
        """
        client = BacklogClient("your_space_name", "your_api_key")
        client.delete_version("YOUR_PROJECT", 3)
        """
        return self.do("DELETE", "projects/{project_id_or_key}/versions/{version_id}",
                       url_params={"project_id_or_key": project_id_or_key,
                                   "version_id": version_id},
                       )

    def issues(self, extra_query_params={}):
        """
        client = BacklogClient("your_space_name", "your_api_key")
        client.issues()

        project_id = client.get_project_id("YOUR_PROJECT")
        client.issues({"projectId[]":[project_id], "sort": "dueDate"})
        """
        return self.do("GET", "issues", query_params=extra_query_params)

    def issue(self, issue_id_or_key):
        """
        client = BacklogClient("your_space_name", "your_api_key")
        client.issue("YOUR_PROJECT-999")
        """
        return self.do("GET", "issues/{issue_id_or_key}",
                       url_params={"issue_id_or_key": issue_id_or_key},
                       )

    def issue_comments(self, issue_id_or_key, extra_query_params={}):
        """
        client = BacklogClient("your_space_name", "your_api_key")
        client.issue_comments("YOUR_PROJECT-999")
        """
        return self.do("GET", "issues/{issue_id_or_key}/comments",
                        url_params={"issue_id_or_key": issue_id_or_key},
                        query_params=extra_query_params
                        )

    def project_issue_types(self, project_id_or_key):
        """
        client = BacklogClient("your_space_name", "your_api_key")
        client.project_issue_types("YOUR_PROJECT")
        """
        return self.do("GET", "projects/{project_id_or_key}/issueTypes",
                       url_params={"project_id_or_key": project_id_or_key},
                       )

    def priorities(self):
        """
        client = BacklogClient("your_space_name", "your_api_key")
        client.priorities()
        """
        return self.do("GET", "priorities")

    def create_issue(self, project_id, summary, issue_type_id, priority_id, extra_request_params={}):
        """
        client = BacklogClient("your_space_name", "your_api_key")
        project_key = "YOUR_PROJECT"

        project_id = client.get_project_id(project_key)
        issue_type_id = client.project_issue_types(project_key)[0][u"id"]
        priority_id = client.priorities()[0][u"id"]

        client.create_issue(project_id,
                            u"some summary",
                            issue_type_id,
                            priority_id,
                            {"description": u"a is b and c or d."})
        """
        request_params = extra_request_params
        request_params["projectId"] = project_id
        request_params["summary"] = summary
        request_params["issueTypeId"] = issue_type_id
        request_params["priorityId"] = priority_id

        return self.do("POST", "issues",
                       request_params=request_params,
                       )

    def add_issue_comment(self, issue_id_or_key, content, extra_request_params={}):
        """
        client = BacklogClient("your_space_name", "your_api_key")
        client.add_issue_comment("YOUR_PROJECT-999", u"or ... else e.")
        """
        request_params = extra_request_params
        request_params["content"] = content
        return self.do("POST", "issues/{issue_id_or_key}/comments",
                       url_params={"issue_id_or_key": issue_id_or_key},
                       request_params=request_params,
                       )

    def users(self):
        """
        client = BacklogClient("your_space_name", "your_api_key")
        client.users()
        """
        return self.do("GET", "users")

    def user(self, user_id):
        """
        client = BacklogClient("your_space_name", "your_api_key")
        client.user(3)
        """
        return self.do("GET", "users/{user_id}", url_params={"user_id": user_id})

    def user_activities(self, user_id, extra_query_params={}):
        """
        client = BacklogClient("your_space_name", "your_api_key")
        client.user_activities(3)
        client.user_activities(3, {"count": 2, "order": "asc"})
        """
        return self.do("GET", "users/{user_id}/activities",
                       url_params={"user_id": user_id},
                       query_params=extra_query_params)

    def groups(self, extra_query_params={}):
        """
        client = BacklogClient("your_space_name", "your_api_key")
        client.groups()
        """
        return self.do("GET", "groups", query_params=extra_query_params)

    def group(self, group_id):
        """
        client = BacklogClient("your_space_name", "your_api_key")
        client.group(3)
        """
        return self.do("GET", "groups/{group_id}", url_params={"group_id": group_id})

    def user_stars(self, user_id, extra_query_params={}):
        """
        client = BacklogClient("your_space_name", "your_api_key")
        client.user_stars(5)
        client.user_stars(5, {"count": 100, "order": "asc"})
        """
        return self.do("GET", "users/{user_id}/stars",
                       url_params={"user_id": user_id},
                       query_params=extra_query_params)

    def user_stars_count(self, user_id, extra_query_params={}):
        """
        client = BacklogClient("your_space_name", "your_api_key")
        client.user_stars_count(5)
        client.user_stars_count(5, {"since": "2017-05-01", "until": "2017-05-31"})
        """
        return self.do("GET", "users/{user_id}/stars/count",
                       url_params={"user_id": user_id},
                       query_params=extra_query_params)

    def star(self, query_params):
        """
        client = BacklogClient("your_space_name", "your_api_key")
        client.star({"issueId": 333})
        """
        return self.do("POST", "stars",
                       query_params=query_params)

    def wikis(self, project_id_or_key):
        """
        client = BacklogClient("your_space_name", "your_api_key")
        client.wikis(3)
        """
        return self.do("GET", "wikis",
                       query_params={"projectIdOrKey": project_id_or_key})

    def wiki(self, wiki_id):
        """
        client = BacklogClient("your_space_name", "your_api_key")
        client.wiki(3)
        """
        return self.do("GET", "wikis/{wiki_id}", url_params={"wiki_id": wiki_id})

    def update_wiki(self, wiki_id, extra_request_params={}):
        """
        client = BacklogClient("your_space_name", "your_api_key")
        client.update_wiki(3, {"name": "test", "content": "content test", "mailNotify": "true"})
        """
        request_params = extra_request_params
        return self.do("PATCH", "wikis/{wiki_id}", url_params={"wiki_id": wiki_id}, request_params=request_params)
                       
    def wiki_history(self, wiki_id):
        """
        client = BacklogClient("your_space_name", "your_api_key")
        client.wiki_history(3)
        """
        return self.do("GET", "wikis/{wiki_id}/history", url_params={"wiki_id": wiki_id})

    def wiki_stars(self, wiki_id):
        """
        client = BacklogClient("your_space_name", "your_api_key")
        client.wiki_stars(3)
        """
        return self.do("GET", "wikis/{wiki_id}/stars", url_params={"wiki_id": wiki_id})

    def project_statuses(self, project_id_or_key):
        """
        client = BacklogClient("your_space_name", "your_api_key")
        client.project_statuses("YOUR_PROJECT")
        """
        return self.do("GET", "projects/{project_id_or_key}/statuses",
                       url_params={"project_id_or_key": project_id_or_key},
                       )

    # -------------------------------
    # extra utilities (PR welcome)
    # -------------------------------

    def get_project_id(self, project_key_or_name):
        projects = self.projects()
        for p in projects:
            if p[u"projectKey"] == project_key_or_name:
                return p[u"id"]
        for p in projects:
            if p[u"name"] == project_key_or_name:
                return p[u"id"]
        return None

    def get_issue_id(self, issue_key):
        issue = self.issue(issue_key)
        if issue:
            return int(issue[u"id"])
        return None
